### PR TITLE
Processing time series

### PR DESCRIPTION
**Pull request by ckittl**
Finally closing #73. This pull request offers possibilities to process time series.

### CODE REVIEW

**Comment by ckittl** [comment]
Comment regarding the Codacy issue: I would like to discuss, if the field names should appear in camel or in snake case. I actually don't know if we agreed on something, yet, but previously they have been snake case. Maybe you @mia-krause can share some experience on best pratice?

**Comment by ckittl** [comment]
> Comment regarding the Codacy issue (unused method): I would like to discuss, if the field names should appear in camel or in snake case. I actually don't know if we agreed on something, yet, but previously they have been snake case. Maybe you @mia-krause can share some experience on best practice?

As of discussion with @johanneshiry and @sensarmad we will do the conversion from camel to snake case in the sinks and sources, as we expect this to be unique for csv files.

File: src/main/java/edu/ie3/datamodel/io/processor/ProcessorProvider.java [hunk 1]
```diff
@@ -143,8 +251,29 @@ public ProcessorProvider(Collection<EntityProcessor<? extends UniqueEntity>> pro
     return resultingProcessors;
   }
 
+  /**
+   * Create processors for all known eligible combinations and map them
+   *
```
Comment (sensarmad) [inline src/main/java/edu/ie3/datamodel/io/processor/ProcessorProvider.java:256]: I think I understand why this mapping is needed (guess also had a discussion) but why isn't it needed for entity processors? easier to ask: what is the difference between time series and entity processors?

File: src/main/java/edu/ie3/datamodel/io/processor/timeseries/TimeSeriesProcessor.java [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+  }
+
+  @Override
+  protected List<Class<? extends TimeSeries>> getEligibleEntityClasses() {
+    return eligibleKeys.stream()
+        .map(TimeSeriesProcessorKey::getTimeSeriesClass)
+        .distinct()
+        .collect(Collectors.toList());
+  }
+}
```
Comment (sensarmad) [inline src/main/java/edu/ie3/datamodel/io/processor/timeseries/TimeSeriesProcessor.java:313]: again, understand what is happening more or less but would like to get a clearer vision

File: src/main/java/edu/ie3/datamodel/models/timeseries/TimeSeries.java [hunk 3]
```diff
@@ -6,36 +6,34 @@
...
-abstract class TimeSeries<T extends Value> extends UniqueEntity {
+/**
+ * Describes a Series of {@link edu.ie3.datamodel.models.value.Value values}
+ *
+ * @param <E> Type of the entries, the time series is foreseen to contain
+ * @param <V> Type of the values, the entries will have
+ */
+public abstract class TimeSeries<E extends TimeSeriesEntry<V>, V extends Value>
+    extends UniqueEntity {
+  protected final Set<E> entries;
```
Comment (sensarmad) [inline src/main/java/edu/ie3/datamodel/models/timeseries/TimeSeries.java:22]: can a time series have different type of entries?

File: src/main/java/edu/ie3/datamodel/models/timeseries/individual/IndividualTimeSeries.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+  /**
+   * Returns the sorted set of all entries known to this time series
+   *
+   * @return An unmodifiable sorted set of all known time based values of this time series
+   */
+  @Override
+  public SortedSet<TimeBasedValue<V>> getEntries() {
+    TreeSet<TimeBasedValue<V>> sortedEntries = new TreeSet<>(timeToValue.values());
+    return Collections.unmodifiableSortedSet(sortedEntries);
```
Comment (sensarmad) [inline src/main/java/edu/ie3/datamodel/models/timeseries/individual/IndividualTimeSeries.java:35]: why are we sorting a time series?

File: src/main/java/edu/ie3/datamodel/io/CsvFileDefinition.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  protected static final String FILE_EXTENSION = "csv";
+
+  protected final String fileName;
+  protected final String[] headLineElements;
+  protected final String csvSep;
+
+  public CsvFileDefinition(String fileName, String[] headLineElements, String csvSep) {
+    if (fileName.matches(fullPathPattern.pattern())) {
+      Matcher matcher = fullPathPattern.matcher(fileName);
+      matcher.matches();
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/io/CsvFileDefinition.java:28]: valid call? returning boolean is ignored ...

File: src/main/java/edu/ie3/datamodel/io/processor/timeseries/TimeSeriesProcessor.java [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+    jointMapping.putAll(valueMapping);
+
+    /* Let uuid be the first entry */
+    return Collections.unmodifiableSortedMap(putUuidFirst(jointMapping));
+  }
+
+  @Override
+  public Optional<LinkedHashMap<String, String>> handleEntity(TimeSeries entity) {
+    throw new UnsupportedOperationException(
+        "Don't invoke this simple method, but TimeSeriesProcessor#handleTimeSeries(TimeSeries).");
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/io/processor/timeseries/TimeSeriesProcessor.java:184]: Okay this should definitely not be implemented that way!  When implemented this way, this makes the whole interface incompatible with all other implementations. 

Example: Someone writes a class that can handle EntityProcessors and now, by default calls `handleEntity` this will cause this person a lot of headaches ...   😨

File: src/main/java/edu/ie3/datamodel/io/sink/DataSink.java [hunk 7]
```diff
@@ -63,18 +66,13 @@
...
+   * extracted and persisted accordingly.
    *
-   * @param entities the entities that should be persisted
-   * @param <C> bounded to be all unique entities. Handling of specific entities is normally then
-   *     executed by a specific {@link EntityProcessor}
+   * @param timeSeries Time series to persist
+   * @param <E> Type of entry in the time series
+   * @param <V> Type of actual value, that is inside the entry
    */
-  <C extends UniqueEntity> void persistAllIgnoreNested(Collection<C> entities);
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/io/sink/DataSink.java:79]: Why did you remove this interface method?

File: src/main/java/edu/ie3/datamodel/models/UniqueEntity.java [hunk 8]
```diff
@@ -10,7 +10,10 @@
 
 /** Supplies every subclass with a generated UUID, making it unique */
 public abstract class UniqueEntity {
-  private final UUID uuid;
+  /** Field name of {@link UniqueEntity} uuid */
+  public static final String UUID_FIELD_NAME = "uuid";
+
+  protected final UUID uuid;
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/models/UniqueEntity.java:16]: image
```suggestion
  private final UUID uuid;
```

File: src/main/java/edu/ie3/datamodel/models/value/WeatherValue.java [hunk 9]
```diff
@@ -15,13 +15,13 @@
 /** Describes weather as a combination of irradiation, temperature and wind values */
 public class WeatherValue implements Value {
   /** The coordinate of this weather value set */
-  private Point coordinate;
+  private final Point coordinate;
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/models/value/WeatherValue.java:18]: :+1: for `final`

File: src/test/groovy/edu/ie3/datamodel/io/FileNamingStrategyTest.groovy [hunk 10]
```diff
@@ -53,20 +54,114 @@ import edu.ie3.datamodel.models.result.system.LoadResult
...
-		FileNamingStrategy strategy = new FileNamingStrategy()
+	def "The FileNamingStrategy is capable of cleaning up strings correctly"() {
+		when:
+		String actual = FileNamingStrategy.cleanString(input)
+
+		then:
+		actual == expected
+
+		where:
+		input 		|| expected
```
Comment (johanneshiry) [inline src/test/groovy/edu/ie3/datamodel/io/FileNamingStrategyTest.groovy:72]: 🤯

File: src/test/groovy/edu/ie3/datamodel/models/timeseries/IndividualTimeSeriesSpec.groovy [hunk 11]
```diff
@@ -5,35 +5,25 @@
...
-	IndividualTimeSeries<IntValue> timeSeries = new IndividualTimeSeries<>(
-	UUID.randomUUID(),
-	new HashMap<ZonedDateTime, IntValue>() { {
-			put(ZonedDateTime.of(1990, 1, 1, 0, 0, 0, 0, ZoneId.of("UTC")), new IntValue(3))
-			put(ZonedDateTime.of(1990, 1, 1, 0, 15, 0, 0, ZoneId.of("UTC")), new IntValue(4))
-			put(ZonedDateTime.of(1990, 1, 1, 0, 30, 0, 0, ZoneId.of("UTC")), new IntValue(1))
-		}
-	})
-
+class IndividualTimeSeriesSpec extends Specification implements TimeSeriesTestData {
```
Comment (johanneshiry) [inline src/test/groovy/edu/ie3/datamodel/models/timeseries/IndividualTimeSeriesSpec.groovy:15]:
```suggestion
class IndividualTimeSeriesTest extends Specification implements TimeSeriesTestData {
```

File: src/test/groovy/edu/ie3/test/helper/UuidPrinter.groovy [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+ * © 2020. TU Dortmund University,
+ * Institute of Energy Systems, Energy Efficiency and Energy Economics,
+ * Research group Distribution grid planning and operation
+ */
+package edu.ie3.test.helper
+
+/**
+ * Simple class to print a bunch of uuids for later static usage in code
+ */
+class UuidPrinter {
```
Comment (johanneshiry) [inline src/test/groovy/edu/ie3/test/helper/UuidPrinter.groovy:11]: image

File: src/main/java/edu/ie3/datamodel/io/CsvFileDefinition.java [hunk 13]
```diff
@@ -23,10 +23,9 @@
   protected final String csvSep;
 
   public CsvFileDefinition(String fileName, String[] headLineElements, String csvSep) {
-    if (fileName.matches(fullPathPattern.pattern())) {
-      Matcher matcher = fullPathPattern.matcher(fileName);
-      matcher.matches();
-      this.fileName = matcher.group(0).replaceAll("\\\\/", File.separator);
+    Matcher fullPathMatcher = fullPathPattern.matcher(fileName);
+    if (fullPathMatcher.matches()) {
```
Comment (johanneshiry) [inline src/main/java/edu/ie3/datamodel/io/CsvFileDefinition.java:27]: 👍

File: src/main/java/edu/ie3/datamodel/io/processor/timeseries/TimeSeriesProcessor.java (hunk 6 above)
Comment (johanneshiry) [reply src/main/java/edu/ie3/datamodel/io/processor/timeseries/TimeSeriesProcessor.java:184]: just dived a little bit more into this and I somehow doubt that the `Processor` in it's current implementation is capable of doing what we want to achieve here due to the differences in output (`Set<LinkedHashMap<String,String>>` vs. `LinkedHashMap<String,String>`). Im wondering if this is somehow solvable in the current concept or if there is no other way to deal with this ....🤔

File: src/main/java/edu/ie3/datamodel/io/sink/DataSink.java (hunk 7 above)
Comment (ckittl) [reply src/main/java/edu/ie3/datamodel/io/sink/DataSink.java:79]: Wasn't called at all.

**Comment by ckittl** [comment]
**Important remark to the merging engineer:**

When ie3-institute/PowerSystemUtils#24 is merged, the class `StringUtils` and the corresponding test can be removed.

**Comment by johanneshiry** [comment]
LGTM - I think we can merge this as soon as this PR is in sync with master :-)
